core/domain/state_domain.py

```diff
@@ -714,6 +714,190 @@ def validate(self):
+
+    def validate(self):
+        """Validates properties of the ContentTranslations.
+
+        Raises:
+            ValidationError: One or more attributes of the ContentTranslations
+            are invalid.
+        """
+        for (content_id, language_code_to_translation_script) in (
+                self.content_translations.iteritems()):
+            if not isinstance(content_id, basestring):
+                raise utils.ValidationError(
+                    'Expected content_id to be a string, received %s'
+                    % content_id)
+            if not isinstance(language_code_to_translation_script, dict):
+                raise utils.ValidationError(
+                    'Expected content_id value to be a dict, received %s'
+                    % language_code_to_translation_script)
+            for (language_code, translation_script) in (
+                    language_code_to_translation_script.iteritems()):
+                if not isinstance(language_code, basestring):
+                    raise utils.ValidationError(
+                        'Expected language_code to be a string, received %s'
+                        % language_code)
+                if not utils.is_valid_language_code(language_code):
+                    raise utils.ValidationError(
+                        'Invalid language_code: %s' % language_code)
+                # TODO(DubeySandeep): Remove this check once App Engine supports
+                # 3-letter language codes in search.
+                if len(language_code) != 2:
+                    raise utils.ValidationError(
+                        'Invalid language_code, it should have exactly 2 '
+                        'letters: %s' % language_code)
+                translation_script.validate()
+
+    def get_available_languages(self):
+        """Returns a set of language available in the ContentTranslations.
+
+        Returns:
+            set. A set of languages available in the ContentTranslations.
+        """
+        languages_list = []
+        for (_, language_code_to_translation_script) in (
+                self.content_translations.iteritems()):
+            for (language_code, _) in (
+                    language_code_to_translation_script.iteritems()):
+                languages_list.append(language_code)
+        return set(languages_list)
+
+    def add_content_id_for_translation(self, content_id):
+        """Adds a content id as a key for the translation into the
+        content_translation dict.
+
+        Args:
+            content_id: str. The id representing a subtitled html.
+
+        Raises:
+            Exception: The content id isn't a string.
+        """
+        if not isinstance(content_id, basestring):
+            raise Exception(
+                'Expected content_id to be a string, received %s' % content_id)
+        if content_id not in self.content_translations:
+            self.content_translations[content_id] = {}
```

Comment:
Should we raise an exception if the content_id already exists? Otherwise this overwrites.

Make sure to include appropriate backend tests.

Reply:
Done!